source/geometry_model/initial_topography_model/ascii_data.cc

```diff
@@ -58,25 +58,32 @@ namespace aspect
     double
     AsciiData<dim>::value (const Point<dim-1> &surface_point) const
     {
-      // Because the get_data_component function of AsciiDataBoundary
-      // expects a dim-dimensional cartesian point, we have to
-      // add a coordinate here, and, for spherical geometries,
-      // change to cartesian coordinates.
+      // In a first step, create a global 'dim'-dimensional point that we can pass to the
+      // function expression as input -- because the function is a dim-dimensional
+      // function.
+      //
+      // Different geometry models pass the surface point in in different ways.
+      // In the following, we will first normalize the input to a dim-dimensional
+      // point with a dummy vertical/radial coordinate that, one would hope,
+      // the AsciiDataBoundary class will then simply ignore.
       Point<dim> global_point;
       if (Plugins::plugin_type_matches<const GeometryModel::Box<dim>> (this->get_geometry_model()) ||
           Plugins::plugin_type_matches<const GeometryModel::TwoMergedBoxes<dim>> (this->get_geometry_model()))
         {
-          // No need to set the vertical coordinate correctly,
-          // because it will be thrown away in get_data_component anyway
           for (unsigned int d=0; d<dim-1; ++d)
             global_point[d] = surface_point[d];
+
+          // Now for the vertical component:
+          global_point[dim-1] = 0;
         }
       else if (Plugins::plugin_type_matches<const GeometryModel::Sphere<dim>> (this->get_geometry_model()) ||
                Plugins::plugin_type_matches<const GeometryModel::SphericalShell<dim>> (this->get_geometry_model()) ||
                Plugins::plugin_type_matches<const GeometryModel::Chunk<dim>> (this->get_geometry_model()))
         {
-          // No need to set the radial coordinate correctly,
-          // because it will be thrown away in get_data_component anyway
+          // TODO: The code here assumes that the AsciiDataBoundary object wants
```

Comment:
I checked, and I think the code is correct, but confusing. AsciiDataBoundary always expects to get the input parameters for its functions in cartesian coordinates. The `get_data_component` function then changes the coordinate system, or more precisely it asks the geometry model to convert the point into its natural coordinate system, before doing the table lookup. Usually this is the most intuitive way to interact with AsciiDataBoundary, because most places in the code work with cartesian coordinates. For some reason the initial topography works differently and already receives coordinates in the natural coordinate system of the geometry models (maybe because it is called from inside the manifold? I dont quite remember). Therefore, here we have to awkwardly convert the point back to cartesian, so that AsciiDataBoundary can do its usual conversion to spherical again if necessary. 
Ideas for improving this process are welcome :-) At a minimum this reasoning should have been documented here.

Reply:
Yes, this is pretty inefficient. We're going around from Cartesian to spherical in the geometry model, then back to Cartesian here, and in the lookup function back up to spherical. 

Thanks for figuring this out. I've put that into the comment. As for a long-term fix: There are not all that many places where the functions in this subsystem are called and processes. It wouldn't be terribly difficult to get those all changed if that's what we want to do. But that's for a separate PR.

Reply:
Ideally, I think we shouldn't be passing around a Cartesian point with a dummy radius, but something that only describes the *surface* point with `dim-1` components.